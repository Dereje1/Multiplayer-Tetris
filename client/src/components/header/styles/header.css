#header {
  width: 100%;
  height: 80px;
  background: #cfd6cf;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
      -ms-flex-pack: center;
          justify-content: center;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
}

#header #login {
  width: 232px;
  height: 70%;
  background: url("btn_google_signin_dark_normal_web@2x.png");
  background-size: 100% 100%;
  background-repeat: no-repeat;
  cursor: pointer;
  outline: none;
}

#header #logout {
  width: 232px;
  height: 70%;
  font-size: 1.2em;
  cursor: pointer;
  outline: none;
}
/*# sourceMappingURL=header.css.map */