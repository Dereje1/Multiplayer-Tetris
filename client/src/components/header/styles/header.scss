#header{
    width: 100%;
    height: 80px;
    background: rgb(207, 214, 207);
    display: flex;
    justify-content: center;
    align-items: center;
    #login{
        width: 232px;
        height: 70%;
        background: url('btn_google_signin_dark_normal_web@2x.png');
        background-size: 100% 100%;
        background-repeat: no-repeat;
        cursor: pointer;
        outline: none;
    }
    #logout{
        width: 232px;
        height: 70%;
        font-size: 1.2em;
        cursor: pointer;
        outline: none;
    }
}