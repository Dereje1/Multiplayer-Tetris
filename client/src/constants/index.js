module.exports = {
  auth: {
    GET_LOGIN_STATUS: 'GET_LOGIN_STATUS',
  },
  game: {
    INITIALIZE_GAME: 'INITIALIZE_GAME',
    LEVEL_UP: 'LEVEL_UP',
    PAUSE: 'PAUSE',
    SET_NEXT_SHAPE: 'SET_NEXT_SHAPE',
    SCREEN_UPDATE: 'SCREEN_UPDATE',
    RAISE_FLOOR: 'RAISE_FLOOR',
    COLLISION: 'COLLISION',
  },
  socket: {
    clientEmit: {
      SEND_LOGGED_IN_USER: 'SEND_LOGGED_IN_USER',
      USER_LOGGED_OUT: 'USER_LOGGED_OUT',
      LOOK_FOR_OPPONENTS: 'LOOK_FOR_OPPONENTS',
      OPPONENT_UNMOUNTED: 'OPPONENT_UNMOUNTED',
      INVITATION_SENT: 'INVITATION_SENT',
      INVITATION_DECLINED: 'INVITATION_DECLINED',
      INVITATION_ACCEPTED: 'INVITATION_ACCEPTED',
      START_GAME: 'START_GAME',
      UPDATED_CLIENT_SCREEN: 'UPDATED_CLIENT_SCREEN',
      GAME_OVER: 'GAME_OVER',
    },
    serverEmit: {
      SERVER_RESET: 'SERVER_RESET',
      LOGGED_IN_USERS: 'LOGGED_IN_USERS',
      SOCKET_ID: 'SOCKET_ID',
      OPPONENT_POOL: 'OPPONENT_POOL',
      UNMOUNT_OPPONENT: 'UNMOUNT_OPPONENT',
      INVITE_SENT: 'INVITE_SENT',
      INVITE_RECIEVED: 'INVITE_RECIEVED',
      DECLINED_INVITATION: 'DECLINED_INVITATION',
      ACCEPTED_INVITATION: 'ACCEPTED_INVITATION',
      GAME_STARTED: 'GAME_STARTED',
      OPPONENT_SCREEN: 'OPPONENT_SCREEN',
      FINISH_GAME: 'FINISH_GAME',
    },
    GAME_COUNTDOWN: 'GAME_COUNTDOWN',
    connection: 'http://localhost:5000/',
  },
};
