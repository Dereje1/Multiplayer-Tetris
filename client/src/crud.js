import axios from 'axios';

const RESTcall = async ({ address, method = 'get', payload = null }) => {
    try {
        const { data } = await axios[method](address, payload);
        return data;
    } catch (error) {
        throw (error);
    }
};

export default RESTcall;