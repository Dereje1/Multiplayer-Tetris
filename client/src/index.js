/* default CRA imports */
import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
/* custom imports */
import { Provider } from 'react-redux';
import store from './redux/store';
import Router from './routes/index';


ReactDOM.render(
  <Provider store={store}>
    <Router />
  </Provider>,
  document.getElementById('root'),
);
