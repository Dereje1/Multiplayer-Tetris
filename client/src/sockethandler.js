import io from 'socket.io-client';
import store from './redux/store';
import { socket as socketConstants } from './constants/index';
import {
  getLoggedInUsers, getClientSocketId, getOpponents, removeOpponents,
  sendInvite, receiveInvite, declinedInvitation, acceptedInvitation,
} from './redux/actions/socket';

const socketConnection = io(socketConstants.connection);
const {
  serverEmit: {
    LOGGED_IN_USERS, SOCKET_ID, SERVER_RESET, OPPONENT_POOL,
    UNMOUNT_OPPONENT, INVITE_SENT, INVITE_RECIEVED, DECLINED_INVITATION,
    ACCEPTED_INVITATION,
  },
} = socketConstants;
const { clientEmit: { SEND_LOGGED_IN_USER } } = socketConstants;

const clientEmitter = (event, dataToEmit) => {
  socketConnection.emit(event, dataToEmit);
};

const serverReset = () => {
  // runs whenever there is a new connection.
  // If client has not refreshed then use existing user profile to update server
  const { socket: socketObject, user } = store.getState();
  if (!Object.keys(socketObject).length) return;
  const payloadToSend = user.profile.authenticated ? user.profile : null;
  clientEmitter(
    SEND_LOGGED_IN_USER,
    payloadToSend,
  );
};

socketConnection.on(
  SERVER_RESET,
  () => serverReset(),
);

socketConnection.on(
  LOGGED_IN_USERS,
  data => store.dispatch(getLoggedInUsers(data)),
);

socketConnection.on(
  SOCKET_ID,
  data => store.dispatch(getClientSocketId(data)),
);

socketConnection.on(
  OPPONENT_POOL,
  data => store.dispatch(getOpponents(data)),
);

socketConnection.on(
  UNMOUNT_OPPONENT,
  () => store.dispatch(removeOpponents()),
);

socketConnection.on(
  INVITE_SENT,
  reciever => store.dispatch(sendInvite(reciever)),
);

socketConnection.on(
  INVITE_RECIEVED,
  sender => store.dispatch(receiveInvite(sender)),
);

socketConnection.on(
  DECLINED_INVITATION,
  () => store.dispatch(declinedInvitation()),
);

socketConnection.on(
  ACCEPTED_INVITATION,
  opponentData => store.dispatch(acceptedInvitation(opponentData)),
);

export default clientEmitter;
