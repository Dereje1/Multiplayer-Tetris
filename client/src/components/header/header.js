import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Redirect, withRouter } from 'react-router-dom';
import getUser from '../../redux/actions/authentication';
import { clientEmitter } from '../../sockethandler';
import { socket as socketConstants } from '../../constants/index';
import Menu from '../menu/menu';
import Invitation from './invitation';
import './styles/header.scss';
import soundFile from './styles/invite_recieved.wav';

const mapStateToProps = state => state;
const mapDispatchToProps = dispatch => bindActionCreators({ getUser }, dispatch);
const {
  clientEmit: {
    SEND_LOGGED_IN_USER, USER_LOGGED_OUT, INVITATION_DECLINED,
    LOOK_FOR_OPPONENTS, INVITATION_ACCEPTED,
  },
} = socketConstants;

class Header extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      inviteReceived: [],
      inviteAccepted: null,
    };
    this.audio = React.createRef();
  }

  componentDidMount() {
    this.loadUser();
  }

  componentDidUpdate() {
    if (!this.audio) return;
    if (!this.audio.current) return;
    const { inviteAccepted, inviteReceived } = this.state;
    const { socket } = this.props;
    if (Object.keys(socket).includes('temp')) {
      if (Object.keys(socket.temp).includes('invitationFrom') && (!inviteReceived || !inviteReceived.length)) {
        const { displayname, difficulty } = socket.temp.invitationFrom;
        this.setState({ inviteReceived: [displayname.split(' ')[0], difficulty] });
        this.audio.current.play()
          .then(() => {
            console.log('Played Audio');
          })
          .catch(e => console.log(`error in audio ${e}`));
      }
      if (!Object.keys(socket.temp).includes('invitationFrom') && inviteReceived && inviteReceived.length) {
        this.setState({ inviteReceived: null });
      }
      if (inviteAccepted) this.setState({ inviteAccepted: null });
    } else if (inviteReceived && inviteReceived.length) this.setState({ inviteReceived: null });
  }

  loadUser = async () => {
    const { getUser: getUserStatus } = this.props;
    await getUserStatus();
    const { user } = this.props;
    const payloadToSend = user.profile.authenticated ? user.profile : null;
    clientEmitter(
      SEND_LOGGED_IN_USER,
      payloadToSend,
    );
  }

  logOutUser = async () => {
    const { getUser: getUserStatus } = this.props;
    await getUserStatus();
    const { user } = this.props;
    user.profile.remove = true;
    clientEmitter(USER_LOGGED_OUT, user.profile);
    window.location.assign('/auth/logout');
  }

  declineInvite = () => {
    const { socket: { temp } } = this.props;
    clientEmitter(INVITATION_DECLINED, temp);
    clientEmitter(LOOK_FOR_OPPONENTS, null);
  };

  acceptInvite = () => {
    const { socket } = this.props;
    this.setState({ inviteAccepted: true }, () => clientEmitter(INVITATION_ACCEPTED, socket.temp));
  }

  audioPlayer = () => (
    <audio ref={this.audio} src={soundFile}>
      <track kind="captions" />
    </audio>
  )

  render() {
    const { inviteReceived, inviteAccepted } = this.state;
    const { user, socket } = this.props;
    const usersMessage = socket.usersLoggedIn
      ? `${socket.usersLoggedIn}  logged in user${socket.usersLoggedIn < 2 ? '' : 's'}`
      : null;
    if (user.profile) {
      const { user: { profile: { authenticated } } } = this.props;
      return (
        <div id="header">
          {
            authenticated
              ? (
                <React.Fragment>
                  <Menu onLogOut={this.logOutUser} />
                  <Invitation
                    invite={inviteReceived}
                    onDeclineInvite={this.declineInvite}
                    onAcceptInvite={this.acceptInvite}
                  />
                  {
                    inviteAccepted
                      ? <Redirect to="/" />
                      : null
                  }
                </React.Fragment>
              )
              : <div id="authbutton"><button id="login" type="submit" onClick={() => window.location.assign('/auth/google')} /></div>
          }
          <div
            id="users"
            role="button"
            tabIndex={-1}
            onKeyDown={() => {}}
            onClick={() => clientEmitter('pool', null)}
          >
            <span>{usersMessage}</span>
          </div>
          {this.audioPlayer()}
        </div>
      );
    }
    return <div id="header" />;
  }

}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Header));

Header.defaultProps = {
  user: null,
  socket: null,
};

Header.propTypes = {
  getUser: PropTypes.func.isRequired,
  socket: PropTypes.shape({
    usersLoggedIn: PropTypes.number,
  }),
  user: PropTypes.shape({
    profile: PropTypes.shape({
      authenticated: PropTypes.bool.isRequired,
    }),
  }),
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
  }).isRequired,
};
