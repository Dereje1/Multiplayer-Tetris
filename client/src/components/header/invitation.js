import React from 'react';
import PropTypes from 'prop-types';
import './styles/header.scss';
import {
  DialogTitle, Dialog, Button, ButtonGroup, Chip, Avatar, DialogContent, DialogActions,
} from '@material-ui/core';

const Invitation = ({
  invite, onDeclineInvite, onAcceptInvite,
}) => {
  const isOpen = Boolean(invite) && Boolean(invite.length);
  if (!isOpen) return null;
  return (
    <Dialog aria-labelledby="simple-dialog-title" open={isOpen}>
      <DialogTitle id="simple-dialog-title">{`${invite[0]} invites you to a duel!`}</DialogTitle>

      <DialogContent>
        <Chip
          color="primary"
          avatar={<Avatar>{invite[1]}</Avatar>}
          label="Difficulty"
          size="medium"
          style={{ marginLeft: 70 }}
        />
      </DialogContent>
      <DialogActions>
        <ButtonGroup
          color="primary"
          aria-label="vertical outlined primary button group"
        >
          <Button
            variant="contained"
            size="large"
            onClick={() => onAcceptInvite()}
            style={{ marginLeft: 10, marginBottom: 20, background: 'green' }}
          >
        Accept
          </Button>
          <Button
            variant="contained"
            size="large"
            onClick={() => onDeclineInvite()}
            style={{ marginLeft: 50, marginBottom: 20, background: 'red' }}
          >
        Decline
          </Button>
        </ButtonGroup>
      </DialogActions>

    </Dialog>
  );
};

export default Invitation;

Invitation.defaultProps = {
  invite: null,
};

Invitation.propTypes = {
  invite: PropTypes.arrayOf(PropTypes.any),
  onDeclineInvite: PropTypes.func.isRequired,
  onAcceptInvite: PropTypes.func.isRequired,
};
