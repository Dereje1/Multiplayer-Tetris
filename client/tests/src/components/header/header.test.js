/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Header } from '../../../../src/components/header/header';
import { stubProfile } from '../../../stub';
import * as socketActions from '../../../../src/sockethandler'

const getRefSpy = () => jest.spyOn(React, "createRef").mockImplementation(() => ({
    current: { play: jest.fn(() => Promise.resolve()) }
}))

describe('The header', () => {
    let props;
    let socketSpy;
    let windowSpy;
    let assign = jest.fn();
    beforeEach(() => {
        props = {
            user: {
                profile: {
                    authenticated: false,
                    userip: "::ffff:127.0.0.1",
                    username: null,
                    displayname: null
                }
            },
            socket: { mySocketId: 'stub_my_socket_id', usersLoggedIn: 0 },
            location: { pathname: '/' },
            getUser: jest.fn()
        }
        socketSpy = jest.spyOn(socketActions, 'clientEmitter');
        windowSpy = jest.spyOn(global, "window", "get");
        windowSpy.mockImplementation(() => ({ location: { assign } }));
    })
    afterEach(() => {
        props = null
        socketSpy.mockClear()
        windowSpy.mockClear()
        assign.mockClear()
    })

    test('will render for unauthenticated users', async () => {
        const wrapper = shallow(<Header {...props} />)
        await Promise.resolve()
        expect(socketSpy).toHaveBeenCalledWith("SEND_LOGGED_IN_USER", null)
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    test('will render for authenticated users', async () => {
        const updatedProps = {
            ...props,
            user: { ...stubProfile.user }
        }
        const wrapper = shallow(<Header {...updatedProps} />)
        await Promise.resolve()
        expect(socketSpy).toHaveBeenCalledWith("SEND_LOGGED_IN_USER", {
            authenticated: true,
            userip: "::ffff:127.0.0.1",
            username: '0001',
            displayname: 'Stub Display Name'
        })
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    test('will logout authenticated users', async () => {
        const updatedProps = {
            ...props,
            user: { ...stubProfile.user }
        }
        const wrapper = shallow(<Header {...updatedProps} />)
        //remove login emit
        await Promise.resolve()
        socketSpy.mockClear()
        const logOutButton = wrapper.find('withRouter(Menu)')
        logOutButton.props().onLogOut()
        await Promise.resolve()
        expect(socketSpy).toHaveBeenCalledWith("USER_LOGGED_OUT", {
            authenticated: true,
            userip: "::ffff:127.0.0.1",
            username: '0001',
            displayname: 'Stub Display Name',
            remove: true
        })
        expect(assign).toHaveBeenCalledWith('/auth/logout')
    })

    test('will accept invites', () => {
        const updatedProps = {
            ...props,
            user: { ...stubProfile.user },
            socket: {
                ...props.socket,
                temp: { 'stub_temp_object': {} }
            }
        }
        const wrapper = shallow(<Header {...updatedProps} />, { disableLifecycleMethods: true })
        expect(wrapper.state().inviteAccepted).toBe(null)
        const invitation = wrapper.find('Invitation')
        invitation.props().onAcceptInvite()
        expect(socketSpy).toHaveBeenCalledWith("INVITATION_ACCEPTED", { 'stub_temp_object': {} })
        expect(wrapper.state().inviteAccepted).toBe(true)
    })

    test('will decline invites', () => {
        const updatedProps = {
            ...props,
            user: { ...stubProfile.user },
            socket: {
                ...props.socket,
                temp: { 'stub_temp_object': {} }
            }
        }
        const wrapper = shallow(<Header {...updatedProps} />)
        const invitation = wrapper.find('Invitation')
        invitation.props().onDeclineInvite()
        expect(socketSpy).toHaveBeenNthCalledWith(1, "INVITATION_DECLINED", { 'stub_temp_object': {} })
        expect(socketSpy).toHaveBeenNthCalledWith(2, "LOOK_FOR_OPPONENTS", null)
    })
})

describe('Component updates (CDU)', () => {
    let props;
    let socketSpy;
    let useRefSpy;
    beforeEach(() => {
        props = {
            user: { ...stubProfile.user },
            socket: { mySocketId: 'stub_my_socket_id', usersLoggedIn: 0, temp: { 'stub_temp_object': {} } },
            location: { pathname: '/' },
            getUser: jest.fn()
        }
        socketSpy = jest.spyOn(socketActions, 'clientEmitter');
        useRefSpy = getRefSpy();
    })
    afterEach(() => {
        props = null
        socketSpy.mockClear()
    })

    test('will set the invite recieved state', () => {
        const updatedProps = {
            ...props,
            socket: {
                ...props.socket,
                temp: {
                    invitationFrom: {
                        displayname: 'stub_Invite_from',
                        difficulty: 3
                    }
                }
            }
        }
        const wrapper = shallow(<Header {...updatedProps} />, { disableLifecycleMethods: true })
        wrapper.instance().componentDidUpdate()
        expect(wrapper.state().inviteReceived).toEqual(["stub_Invite_from", 3])
    })

    test('will reset the invite recieved state', () => {
        const wrapper = shallow(<Header {...props} />, { disableLifecycleMethods: true })
        wrapper.setState({ inviteReceived: ['invite_info'] })
        wrapper.instance().componentDidUpdate()
        expect(wrapper.state().inviteReceived).toBe(null)
    })

    test('will reset the invite accepted state', () => {
        const wrapper = shallow(<Header {...props} />, { disableLifecycleMethods: true })
        wrapper.setState({ inviteAccepted: true })
        wrapper.instance().componentDidUpdate()
        expect(wrapper.state().inviteAccepted).toBe(null)
    })
})