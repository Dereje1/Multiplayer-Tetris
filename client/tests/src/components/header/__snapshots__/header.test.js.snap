// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The header will render for authenticated users 1`] = `
<div
  id="header"
>
  <withRouter(Menu)
    onLogOut={[Function]}
  />
  <Invitation
    invite={Array []}
    onAcceptInvite={[Function]}
    onDeclineInvite={[Function]}
  />
  <div
    id="users"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={-1}
  >
    <span />
  </div>
  <audio
    src={
      Object {
        "canInstrument": true,
        "createTransformer": [Function],
        "getCacheKey": [Function],
        "getCacheKeyAsync": [Function],
        "process": [Function],
        "processAsync": [Function],
      }
    }
  >
    <track
      kind="captions"
    />
  </audio>
</div>
`;

exports[`The header will render for unauthenticated users 1`] = `
<div
  id="header"
>
  <div
    id="authbutton"
  >
    <button
      id="login"
      onClick={[Function]}
      type="submit"
    />
  </div>
  <div
    id="users"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={-1}
  >
    <span />
  </div>
  <audio
    src={
      Object {
        "canInstrument": true,
        "createTransformer": [Function],
        "getCacheKey": [Function],
        "getCacheKeyAsync": [Function],
        "process": [Function],
        "processAsync": [Function],
      }
    }
  >
    <track
      kind="captions"
    />
  </audio>
</div>
`;
